server/src/models/user.js

```
@@ -0,0 +1,126 @@
+    trim: true,
+    required: true,
+    minlength: 2,
+  },
+  email: {
+    type: String,
+    required: true,
+    trim: true,
+    minlength: 1,
+    unique: true,
+    validate: {
+      validator: validator.isEmail,
+      message: "{VALUE} is not a valid email",
+    },
+  },
+  password: {
+    type: String,
+    required: true,
+    minlength: 6,
+  },
+  tokens: [
+    {
+      access: {
+        type: String,
+        required: true,
+      },
+      token: {
+        type: String,
+        required: true,
+      },
+    },
+  ],
+});
+
+// instance methods
+// determines what to send back when mongoose model is converted to JSON
+UserSchema.methods.toJSON = function toJSON() {
+  const user = this;
+  const { _id, name, email } = user.toObject();
+  return { _id, name, email };
+};
+
+UserSchema.methods.generateAuthToken = async function generateAuthToken() {
+  const user = this;
+  const access = "auth";
+  // eslint-disable-next-line no-underscore-dangle
+  const _id = user._id.toHexString();
+  const token = jwt.sign({ _id, access }, process.env.JWT_SECRET).toString();
+  user.tokens = [...user.tokens, { access, token }];
+  await user.save();
+  return token;
+};
+
+UserSchema.methods.removeToken = function removeToken(token) {
+  const user = this;
+  return user.update({
+    $pull: {
+      tokens: { token },
+    },
+  });
+};
+
+// model methods
+UserSchema.statics.findByToken = async function findByToken(token) {
+  const User = this;
+  let decoded;
+  try {
+    decoded = jwt.verify(token, process.env.JWT_SECRET);
+  } catch (error) {
+    return Promise.reject();
+  }
+  const { _id } = decoded;
+  const user = await User.findOne({
+    _id,
+    "tokens.token": token,
+    "tokens.access": "auth",
+  });
+  return user;
+};
+
+UserSchema.statics.findByCredentials = async function findByCredentials(
+  email,
+  password
+) {
+  const User = this;
+  const user = await User.findOne({ email });
+  if (!user) {
+    return Promise.reject();
```

Why? you can throw an error simply over here

---

corrected.